### PR TITLE
OSQLStaticReflectiveFunction to automatically bind static methods

### PR DESCRIPTION
**Pull request by fabriziofortino**
* Added mechanism to automatically register static methods of a given class to sql functions
* Registered `java.lang.Math` methods with `math_` prefix

#2573

### CODE REVIEW

**Comment by luigidellaquila** [comment]
Hi @fabriziofortino 

Thank you, I had a quick look at this and I like it very much.
I'll do a full review in next hours and keep you posted

Thanks!

Luigi

**Comment by fabriziofortino** [comment]
Good to hear that @luigidellaquila 

With OSQLStaticReflectiveFunction we could easily register a number of other useful functions (I have some idea in mind). 

Looking forward to your comments.

File: core/src/main/java/com/orientechnologies/orient/core/sql/functions/misc/OSQLStaticReflectiveFunction.java [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+    });
+  }
+
+  @Override
+  public Object execute(Object iThis, OIdentifiable iCurrentRecord, Object iCurrentResult, Object[] iParams,
+      OCommandContext iContext) {
+
+    Method method = pickMethod(iParams);
+
+    try {
```
Comment (luigidellaquila) [inline core/src/main/java/com/orientechnologies/orient/core/sql/functions/misc/OSQLStaticReflectiveFunction.java]: Shouldn't we add a null check here...?

**Comment by luigidellaquila** [comment]
Hi Fabrizio,

I just reviewed the PR and I'd say it's OK. I only added a comment about a null check we could add, just to make it perfect. WDYT?

Thanks

Luigi

**Comment by fabriziofortino** [comment]
@luigidellaquila PR amended to address your feedback

**Comment by luigidellaquila** [comment]
Great, thank you very much, merging...

**Comment by luigidellaquila** [comment]
Merged, thank you very much @fabriziofortino !

Luigi

**Comment by lvca** [comment]
One of the best PR ever received! Good job @fabriziofortino !

**Comment by luigidellaquila** [comment]
I was wondering if we can use this mechanism as a simplified way for end users to create custom functions.
Any Java developer can write a class with a static method and create a Jar file. A small additional configuration (perhaps in the orientdb-server-config.xml) to define the function prefix and it's done!
WDYT?

Thanks

Luigi

**Comment by fabriziofortino** [comment]
I was actually thinking to use the same mechanism to bind the string similarity and distance algorithms provided by https://github.com/tdebatty/java-string-similarity/ (It's a very small library ... around 45KB). I think these functions can be very useful in a database. Oracle has something similar too (https://docs.oracle.com/cd/E18283_01/appdev.112/e16760/u_match.htm).

With that said I could try to implement the generic mechanism @luigidellaquila proposed and use java-string-similarity to prove it works. What about creating a separate github repo that contains all the **EXTRA** functions?

@luigidellaquila @lvca WDYT?

**Comment by luigidellaquila** [comment]
Hi @fabriziofortino 

For me the idea of having a separate project for extra functions is very good, it will make community contributions much easier.
@lvca wdyt?

Thanks

Luigi

**Comment by lvca** [comment]
I like it. The best would be creating multiple directories by type, so when we compile we generate multiple jars, and the user can pick whatever he wants. WDYT?

**Comment by fabriziofortino** [comment]
I will create a separate PR to allow external custom functions as @luigidellaquila suggested.

After this we can work on the extra function repo.
Do you want to create it under the `orientechnologies` github organization?

**Comment by lvca** [comment]
Just created: https://github.com/orientechnologies/extra-functions

**Pull request merged**